### PR TITLE
Add evictAll to blob cache

### PR DESCRIPTION
**Pull request by JVerwolf**
This PR adds the changes needed for adding/testing a clear blob-cache API to serverless.

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -623,6 +629,7 @@ public void forceEvict(Predicate<KeyType> cacheKeyPredicate) {
                 }
             }
         }
+        return matchingEntries.size();
```
Comment (JVerwolf) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:632]: When performing an side-effecting (mutation of state) operation, it's nice to provide _feedback_ when something is done. This comes from the idea of feedback in UX design. Even if we don't use this value, it can be used to show a  (human) callee that an effect took place.

Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:632]: We should be counting only when `evicted` is true

**Review by tlrx — COMMENTED** [review]
I made a comment about how evictions should be counted. Also, can we remove the other changes that seem unrelated?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java (hunk 1 above)
Comment (JVerwolf) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:632]: Right, good catch 👍

**Comment by JVerwolf** [comment]
>  Also, can we remove the other changes that seem unrelated?

These are needed to verify serialization of failures in node response objects, and are part of the same external work that also uses `evictAll()` that this PR adds to the blob cache. If required, I can pull these into a separate PR.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 2]
```diff
@@ -605,24 +605,42 @@ public void removeFromCache(KeyType cacheKey) {
...
         }
+        return evictedCount;
+    }
+
+    /**
+     * Evicts all entries from the cache.
+     *
+     * @return The number of entries evicted from the keyMapping.
+     */
+    public int forceEvictAll() {
```
Comment (javanna) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:642]: Would it be doable to add a unit test for this new method? And possibly update existing tests for forceEvict to verify the returned value?

Side note, for my curiosity: do we need a new public method? We could also easily provide an always true predicate?

File: server/src/main/java/org/elasticsearch/action/FailedNodeException.java [hunk 3]
```diff
@@ -43,4 +44,25 @@ protected void writeTo(StreamOutput out, Writer<Throwable> nestedExceptionsWrite
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(
+            nodeId,
+            getMessage(),
+            getClass().getName(),
+            getStackTrace().length // Could check elements individually, though there can be slight differences when serialized.
+        );
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/FailedNodeException.java:66]: I am wondering if these changes should go to their own PR for visibility, as they have little to do with evicting all items from the blob cache. (I do remember why you needed it in the first place, and a separate PR would highlight the reson why you need it and solicit targeted feedback for that problem.)

**Review by javanna — COMMENTED** [review]
I left two small comments, thanks!

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java (hunk 2 above)
Comment (JVerwolf) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:642]: > do we need a new public method? We could also easily provide an always true predicate?

I've gone back and forth on this, see  https://github.com/elastic/elasticsearch-serverless/pull/462#discussion_r1259499244

Unfortunately, it seems the only location where `forceEvict()` is called from a test is here: https://github.com/JVerwolf/elasticsearch/blob/989b5cfb23309d75012b1f83ddc6368c67927405/x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java#L278
https://github.com/elastic/elasticsearch/blob/989b5cfb23309d75012b1f83ddc6368c67927405/x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java#L278 

I'm currently thinking of reverting this to use `forceEvict()`, as adding a new method to this cache requires doing the  due-diligence of exhaustively testing it, probably in multi-threaded scenarios. If so, this adds more overhead to this PR than I think is worth it given the nature of these changes. WDYT?

Comment (javanna) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:642]: I agree. if the existing tests already offer enough coverage, that seems like a pragmatic solution.

**Comment by JVerwolf** [comment]
As per PR feedback, I've split up the changes. See: https://github.com/elastic/elasticsearch/pull/97124
